cli/src/commands/extensions/api.rs

```diff
@@ -50,6 +49,23 @@ impl<T: Unpin> OnceFuture<T> {
     }
 }
 
+// XXX: Holding a mutable reference to any field inside `ExtensionState` across
+// await points, will cause issues when that field is accessed from another
+// extension API method.
+//
+// Accessing the `ExtensionState` is only safe through `ExtensionStateRef`,
+// since that ensures that `OpState` is not accessed through a mutable reference
+// which could be held across await points.
+//
+// When making a field of `ExtensionState` mutably accessible, the mutation
+// should occur internally through SYNCHRONOUS methods (e.g. `with_x(|x| ...)`),
+// so holding a mutable reference is impossible.
+//
+// If a mutable reference MUST be held across await points, like `PhylumApi`,
+// its synchronization should force blocking through an async-safe `Mutex`. This
+// will stall all extension API methods trying to access this field, so ensure
+// the blocking duration is minimal.
```

Comment:
I've discussed this with @andreaphylum in Slack and I think this is a good summary of how to deal with the potential pitfalls here.

Since @kylewillmon hasn't been part of this conversation, I think it would be good to get the first review of this explanation from you, just to ensure this documentation properly explains the underlying issue even without me or @andreaphylum elaborating in this PRs comments.

Reply:
Hm maybe I should clarify the section about what you can do when you **need** mutable references.

The idea with the `Ref` is fine is basically that the `Rc<RefCell<OpState>>` reference we use to access the `ExtensionState` can be held across await points because we never want to mutably access `ExtensionState`, which is also why I pointed out it should only ever be accessed through `ExtensionStateRef` (rather than bypassing it and calling `borrow_mut` or something).

So in that sense `Ref` is always perfectly safe, as long as there never will be mutable access, you can hold and access however many immutable references you want, because that's not an issue.

The problem only arises when you need to access things mutably. In that scenario you should **never** hold a reference to that data point (mutable or immutable) across await points. And methods like `with_x(|x| ...)` are one way you can try to enforce this.

@kylewillmon So I think you got the gist of it appropriately, but the comment should probably be updated when it comes to creating safe mutable APIs. Does this explanation make sense to you and if so, do you have suggestions for how to put this into one paragraph?

Reply:
> So in that sense `Ref` is always perfectly safe, as long as there never will be mutable access, you can hold and access however many immutable references you want, because that's not an issue.

This is where I think you are mistaken. The `RefCell<OpState>` will panic if any code tries to create a `RefMut` while there are any outstanding `Ref` objects. Even if we never need a `RefMut` ourselves for the `ExtensionState` object contained within, the `RefCell` has no way to distinguish that from any `RefMut` created by `deno_core` to mutate other pieces of `OpState`. Therefore, we are risking a panic anytime we hold a `Ref` across an await point since there may be some `deno_core` code that runs in that time.

Hiding the `Ref` inside of `ExtensionStateRef` just hides the clippy lint and prevents us from being alerted about this potential problem in the future.

Reply:
> This is where I think you are mistaken. The RefCell<OpState> will panic if any code tries to create a RefMut while there are any outstanding Ref objects. Even if we never need a RefMut ourselves for the ExtensionState object contained within, the RefCell has no way to distinguish that from any RefMut created by deno_core to mutate other pieces of OpState. Therefore, we are risking a panic anytime we hold a Ref across an await point since there may be some deno_core code that runs in that time.

This is making the assumption that deno core might mutably access OpState while we're awaiting. As far as I know that doesn't happen, but we could just add another layer of `Rc` in between to resolve that. The issue stays the same either way though. I just decided not to add an `Rc` because it seems to work fine right now.

> Hiding the Ref inside of ExtensionStateRef just hides the clippy lint and prevents us from being alerted about this potential problem in the future.

You say that, but even in your PR we hold a ref across an await point. And there's not much we can do about this. That's the entire reason behind my attempt to document it. We cannot avoid this issue and Clippy will never catch all instances of it (some of which might be justified). So we need some way to document things.